Davidkpiano/v5-updates (#119)
statelyai/docs: versioned_docs/version-5/actors.mdx

import ThemedImage from '@theme/ThemedImage';

When you run a state machine, it becomes an actor: a running process that can receive events, send events and change its behavior based on the events it receives, which can cause effects outside of the actor.

> In state machines, actors can be **invoked** or **spawned**. These are essentially the same, with the only difference being how the lifecycle of the actor is controlled.

Reviewer: That’ll probably be because I missed a `the` 😆

> In state machines, actors can be **invoked** or **spawned**. These are essentially the same, with the only difference being how the lifecycle of the actor is controlled.

Reviewer: (Fixed now)

Actor logic is the logical "model" (brain, blueprint, DNA, etc.) of the actor. It describes how the actor should change its behavior when it receives an event. You can create actor logic using **actor logic creators**. The types of actor logic you can create from XState are:

- [State machine logic (`createMachine(...)`)](#TODO)
- [Promise logic (`fromPromise(...)`)](#TODO)
- [Transition logic (`fromTransition(...)`)](#TODO)
- [Observable logic (`fromObservable(...)`)](#TODO)
- [Event observable logic (`fromEventObservable(...)`)](#TODO)
- [Callback logic (`fromCallback(...)`)](#TODO)

In XState, actor logic is defined by an object containing methods like `.transition(...)`, `.getInitialState()`, `.getSnapshot()`, and more. This object tells an interpreter how to update an actor's internal state when it receives an event, as well as which effects to execute (if any).

## Creating actors

You can create an actor, which is a "live" instance of some actor logic, via `interpret(actorLogic, options?)`. The `interpret(...)` function takes the following arguments:

- `actorLogic` - The [actor logic](#TODO) to interpret
- `options` (optional) - Interpreter options

> When you `interpret(...)` actor logic, you implicitly create an [actor system](#TODO) where the created actor is the root actor. Any actors that are spawned from this root actor and its descendents are part of that actor system. The actor must be started by calling `actor.start()`, which will also start the actor system:

Reviewer: ```suggestion
When you `interpret(...)` actor logic, you implicitly create an [actor system](system.mdx) where the created actor is the root actor. Any actors spawned from this root actor and its descendants are part of that actor system. The actor must be started by calling `actor.start()`, which will also start the actor system:
```

```ts
import { interpret } from 'xstate';
import { someActorLogic } from './someActorLogic.ts';

const actor = interpret(someActorLogic);
actor.start();

// Now the actor can receive events
actor.send({ type: 'someEvent' });
```

> You can stop root actors by calling `actor.stop()`. This will also stop the actor system and all actors in that system:

Reviewer: ```suggestion
You can stop root actors by calling `actor.stop()`, which will also stop the actor system and all actors in that system:
```

## Actor snapshots

> When an actor receives an event, its internal state may change. An actor may emit a **snapshot** when a state transition occurs. You can read an actor's snapshot synchronously via `actor.getSnapshot()`. An actor's snapshot is not necessarily the same as its internal state; rather, it is a value that it wants to _share_ with subscribers. For example, a promise actor has an internal state consisting of `data`, `status`, and other internal properties, but only the `data` is emitted as its snapshot.

Reviewer: ```suggestion
When an actor receives an event, its internal state may change. An actor may emit a **snapshot** when a state transition occurs. You can read an actor’s snapshot synchronously via `actor.getSnapshot()`. 

An actor’s snapshot is not necessarily the same as its internal state; instead, it is a value the actor wants to _share_ with subscribers. For example, a promise actor has an internal state consisting of `data`, `status`, and other internal properties, but only the `data` is emitted as its snapshot.
```

```ts
import { fromPromise, interpret } from 'xstate';

const countLogic = fromPromise(async () => {
  const count = await fetchCount();

  return count;
});

const countActor = interpret(countLogic);
countActor.start();

countActor.getSnapshot(); // logs undefined

// After the promise resolves...
countActor.getSnapshot(); // logs 42
```

## Subscribing to actors

You can subscribe to an actor's snapshot values via `actor.subscribe(observer)`. The observer will receive the actor's snapshot value when it is emitted. The observer can be:

- A plain function that receives the latest snapshot

> - Or an observer object whose `.next(snapshot)` method receives the latest snapshot

Reviewer: ```suggestion
- An observer object whose `.next(snapshot)` method receives the latest snapshot
```

```ts
// Observer as a plain function
const subscription = actor.subscribe((snapshot) => {
  console.log(snapshot);
});
```

```ts
// Observer as an object
const subscription = actor.subscribe({
  next(snapshot) {
    console.log(snapshot);
  },
  error(data) {
    // ...
  },
  complete() {
    // ...
  },
});
```

The return value of `actor.subscribe(observer)` is a subscription object that has an `.unsubscribe()` method. You can call `subscription.unsubscribe()` to unsubscribe the observer:

```ts
const subscription = actor.subscribe((snapshot) => {
  /* ... */
});

// Unsubscribe the observer
subscription.unsubscribe();
```

When the actor is stopped, all observers will automatically be unsubscribed.

## Initializing actor state

> You can initialize actor logic at a specific persisted internal state by passing the state in the 2nd `options` argument of `interpret(logic, options)`. This will create an actor that will be started at that persisted state, if the state is compatible with the actor logic:

Reviewer: ```suggestion
You can initialize actor logic at a specific persisted internal state by passing the state in the second `options` argument of `interpret(logic, options)`. If the state is compatible with the actor logic, this will create an actor that will be started at that persisted state:
```

```ts
const persistedState = JSON.parse(localStorage.get('some-persisted-state'));

const actor = interpret(someLogic, {
  // highlight-start
  state: persistedState,
  // highlight-end
});

// Actor will start at persisted state
actor.start();
```

See [persistence](persistence.mdx) for more details.

## Awaiting actors

> You can wait for an actor's snapshot to satisfy a predicate using the `waitFor(actor, predicate, options?)` helper function. The `waitFor(...)` function returns a promise that is:

Reviewer: ```suggestion
You can wait for an actor’s snapshot to satisfy a predicate using the `waitFor(actor, predicate, options?)` helper function. The `waitFor(...)` function returns a promise that is:
```

- Resolved when the emitted snapshot satisfies the `predicate` function
- Resolved immediately if the current snapshot already satisfies the `predicate` function
- Rejected if an error is thrown or the `options.timeout` value is elapsed.

```ts
await waitFor(
  countActor,
  (snapshot) => {
    return snapshot.count >= 100;
  },
  {
    timeout: 10_000, // 10 seconds (10,000 milliseconds)
  },
);
```

## State machine actor logic

> You can describe actor logic as a [state machine](#TODO). Actors created from state machine actor logic can:

Reviewer: ```suggestion
You can describe actor logic as a [state machine](machines.mdx). Actors created from state machine actor logic can:
```

- Receive events
- Send events to other actors
- Invoke/spawn child actors
- Emit snapshots of its state
- Output a value when the machine reaches its top-level final state

```ts
const toggleMachine = createMachine({
  initial: 'inactive',
  states: {
    inactive: {},
    active: {},
  },
});

const toggleActor = interpret(toggleMachine);
toggleActor.subscribe((snapshot) => {
  console.log(snapshot.value);
});
toggleActor.start();
// Logs 'inactive'
toggleActor.send({ type: 'toggle' });
// Logs 'active'
```

## Promise actor logic

> Promise actor logic is logic that is described by an async process that resolves or rejects after some time. Actors created from promise logic ("promise actors") can:

Reviewer: ```suggestion
Promise actor logic is described by an async process that resolves or rejects after some time. Actors created from promise logic (“promise actors”) can:
```

- Emit the resolved value of the promise
- Output the resolved value of the promise

Sending events to promise actors will have no effect.

```ts
const promiseLogic = fromPromise(() => {
  return fetch('...').then((data) => data.json());
});

const promiseActor = interpret(promiseLogic);
promiseActor.subscribe((snapshot) => {
  console.log(snapshot);
});
promiseActor.start();
// Logs undefined

// After promise resolves
// Logs user: { name: ... }
```

## Transition actors

> Transition actor logic is logic that is described by a [transition function](#TODO), which is similar to a [reducer](#TODO). Transition functions are functions that take the current `state` and received `event` object as arguments, and return the next state. Actors created from transition logic ("transition actors") can:

Reviewer: ```suggestion
Transition actor logic is described by a [transition function](#TODO), similar to a [reducer](#TODO). Transition functions take the current `state` and received `event` object as arguments, and return the next state. Actors created from transition logic (“transition actors”) can:
```

- Receive events
- Emit snapshots of its state

```ts
const transitionLogic = fromTransition(
  (state, event) => {
    if (event.type === 'increment') {
      return {
        ...state,
        count: state.count + 1,
      };
    }
    return state;
  },
  { count: 0 },
);

const transitionActor = interpret(transitionLogic);
transitionActor.subscribe((snapshot) => {
  console.log(snapshot);
});
transitionActor.start();
// Logs { count: 0 }

transitionActor.send({ type: 'increment' });
// Logs { count: 1 }
```

## Observable actors

> Observable actor logic is logic that is described by an [observable stream of values](#TODO). Actors created from observable logic ("observable actors") can:

Reviewer: ```suggestion
Observable actor logic is described by an [observable stream of values](#TODO). Actors created from observable logic (“observable actors”) can:
```

- Emit snapshots of its emitted value

Sending events to observable actors will have no effect.

```ts
const secondLogic = fromObservable(() => interval(1000));

const secondActor = interpret(secondLogic);
secondActor.start();
// At every second:
// Logs 0
// Logs 1
// Logs 2
// ...
```

## Event observable actors

> Event observable actor logic is logic that is described by an observable stream of [event objects](#TODO). Actors created from event observable logic ("event observable actors") can:

Reviewer: ```suggestion
Event observable actor logic is described by an observable stream of [event objects](transitions.mdx#event-objects). Actors created from event observable logic (“event observable actors”) can:
```
